Applications/Terminal/Terminal.cpp

```diff
@@ -470,14 +518,25 @@ void Terminal::scroll_up()
 {
     // NOTE: We have to invalidate the cursor first.
     invalidate_cursor();
-    delete m_lines[0];
-    for (word row = 1; row < rows(); ++row)
+    delete m_lines[m_scroll_region_top];
+    for (word row = m_scroll_region_top + 1; row < m_scroll_region_bottom; ++row)
         m_lines[row - 1] = m_lines[row];
-    m_lines[m_rows - 1] = new Line(m_columns);
+    m_lines[m_scroll_region_bottom - 1] = new Line(m_columns);
     ++m_rows_to_scroll_backing_store;
```

Comment:
I don't think the `m_rows_to_scroll_backing_store` optimization will work correctly if you have a scroll region.
At the very least it probably needs to be aware of it, and take some special action.
Currently it will just move everything up to the first scanline (excluding vertical inset.) If there's a scrolling region set up, that definitely won't do the right thing.